internal/uuid/uuid.go

```diff
@@ -0,0 +1,31 @@
+package uuid
+
+import (
+	"crypto/rand"
+	"fmt"
+	"io"
+)
+
+var rander = rand.Reader
+
+// SetRand Use your own randomizer to support unit testing
+func SetRand(r io.Reader) {
+	if r == nil {
+		rander = rand.Reader
+		return
+	}
+
+	rander = r
+}
+
+// New Create a version 4 random UUID
+func New() (string, error) {
```

Comment:
Instead of the global variable and `SetRand()` please change the signature of `New()` to take in the reader (almost exactly like https://github.com/google/uuid/blob/master/version4.go#L34)

Then where we call it in `runtime.go` pass in the `crypto/rand` `rand.Reader`, and for the builtin implementation use a new random reader with https://golang.org/pkg/math/rand/#New and a `Source` https://golang.org/pkg/math/rand/#NewSource configured to use the seed that was provided as the argument.

Reply:
I still do not understand why you want deterministic behaviour of a uuid function which sole purpose is to provide a truly random string (at least for version 4). I also wonder how you can provide a truly random seed as argument to the uuid function from within a policy. A better solution would be, imho, an optional seed where the crypto/rand Reader is used when no seed has been specified. However, I do not know how to implement optional parameters in Go (I am a novice in Go). I prefer to use the crypto/rand module over the math/rand because the quality of the random result will be significantly higher.

Reply:
Thank you for working with us on this @reneklootwijk I appreciate your patience on the issue.

> I still do not understand why you want deterministic behaviour of a uuid function which sole purpose is to provide a truly random string

There are two main reasons. First, and arguably the most (and maybe only) important one is that OPA does not evaluate these in an imperative top-to-bottom way. Just because the `uuid()` call is in an expression a single time in the policy it does not mean that it will be called a single time. That expression _could_ be evaluated any number of times by OPA while evaluating some top level query, to that end the builtin functions need should be implemented to be idempotent. The second reason is to make policies easier to unit test for policy authors. If they can predict the uuid for a unit test they can more easily compare expected and actual results for some query. Its the same reason why almost all languages/frameworks have options to seed "random" API's.

> I also wonder how you can provide a truly random seed as argument to the uuid function from within a policy.

> I prefer to use the crypto/rand module over the math/rand because the quality of the random result will be significantly higher.

Totally valid points. In my mind I was advocating for the argument to the builtin function to act as the actual random number seed, but it doesn't need to be the case to fulfill the requirements  I described above.

What I would propose is that we have the argument, but we don't call it a "seed". Internally use the builtin context cache https://github.com/open-policy-agent/opa/blob/5e6a1385595b7a09ee7f0184ac1168d2b7d48204/topdown/builtins.go#L34 which is provided as an argument to the builtin implementation if you use the newer registration functions rather than the deprecated `RegisterFunctionalBuiltinX` ones https://github.com/open-policy-agent/opa/blob/5e6a1385595b7a09ee7f0184ac1168d2b7d48204/topdown/builtins.go#L41-L52. 

So we use the parameter as the key, with a more optimally generated uuid as the value. Nothing prevents us from using `crypto/rand`'s reader to make a these UUID's. We then would just need to have a step early in the builtin to check the cache if it already has a uuid generated for the parameter, if it does return it, otherwise generate a new one (save it in the cache) and return that.

The new documented definition is more like:
| Built-in | Description |
| ------- |-------------|
| <span class="opa-keep-it-together">`output := uuid.rfc4122(str)`</span> | `output` is a `string` representing a version 4 UUID. For any given `str` the `output` will be consistent throughout a query evaluation. |

End result is that you generate the most random UUID available, and we can have the function be idempotent for a given parameter.

Thoughts?

Reply:
Just to add onto what @patrick-east said, IMO there are two related but different things to consider here. Both have to do with determinism. I mentioned them very briefly in my original comment but didn't go into depth. Luckily @patrick-east went into some detail already so this might just be repetitive (sorry!)

Having a simple function that accepts no arguments and returns a random UUID may be fine at first glance, however, there are some interesting problems that it would create. For example, suppose you wrote the following:

```rego
package x

p = uuid()
```

And then somewhere else you said: `a = data.x.p; b = data.x.p`. In this case, I'd expect `a` and `b` to have the same value. However, if `uuid()` generated a random value on each invocation, that would not be true--this would be a fundamental change to the semantics. Unfortunately it's not even that simple because in this particular case, because of an optimization in the OPA evaluator for complete rules `a` and `b` WOULD have the same value because OPA memoizes complete rule values/assignments. So now the `uuid()` function does not even act the way we wanted at the outset.

The answer to this particular problem is to be explicit in the policy about when we want a new UUID. We can do this by simply extending the function to accept a parameter that identifies the unique value (@patrick-east described this nicely above.)

So that's the first thing. The other thing is that Rego is intended to be portable, meaning you can ship it around and execute it and get the same result out. This is useful for different kinds of policy enforcement (e.g., define the policy once and then apply it in multiple places like in a cluster, in a CI/CD pipeline, offline for audit purposes, etc.) but also for analysis and testing! Since the policies are deterministic (w/ the exception of a special class of built-in function) you can do things like replay decisions and be confident the answer is the same. This gets us back back to the point of being able to _seed_ the uuid function when evaluation is run. I don't have a strong opinion about whether the policy author should be able to seed or not but the evaluator definitely needs some way to do this otherwise what I said before is not possible. Note, we have the same problem with `time.now_ns()`  and `http.send()` though time and uuids are way easier to deal with.